group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/HdrHistogram.java

```diff
@@ -0,0 +1,145 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.coordinator.group.metrics;
+
+import org.HdrHistogram.Histogram;
+import org.HdrHistogram.Recorder;
+import org.HdrHistogram.ValueRecorder;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * <p>A wrapper on top of the HdrHistogram API. It handles writing to the histogram by delegating
+ * to an internal {@link ValueRecorder} implementation, and reading from the histogram by
+ * efficiently implementing the retrieval of up-to-date histogram data.
+ *
+ * <p>Note that all APIs expect a timestamp which is used by the histogram to discard decaying data
+ * and determine when the snapshot from which the histogram metrics are calculated should be
+ * refreshed.
+ */
+public final class HdrHistogram {
+
+    private static final long DEFAULT_MAX_SNAPSHOT_AGE_MS = 1000L;
+
+    /**
+     * The duration (in millis) after which the latest histogram snapshot is considered outdated and
+     * subsequent calls to {@link #latestHistogram(long)} will result in the snapshot being recreated.
+     */
+    private final long maxSnapshotAgeMs;
+
+    /**
+     * The internal HdrHistogram data structure used to:
+     * <ul>
+     *   <li>ingest data;</li>
+     *   <li>get a {@link Histogram} of the latest ingested data;</li>
+     * </ul>
+     */
+    private final Recorder recorder;
+
+    /**
+     * The latest snapshot of the internal HdrHistogram. Automatically updated by
+     * {@link #latestHistogram(long)} if older than {@link #maxSnapshotAgeMs}.
+     */
+    private final AtomicReference<Timestamped<Histogram>> timestampedHistogramSnapshot;
+
+    public HdrHistogram(
+        long highestTrackableValue,
+        int numberOfSignificantValueDigits
+    ) {
+        this(DEFAULT_MAX_SNAPSHOT_AGE_MS, highestTrackableValue, numberOfSignificantValueDigits);
+    }
+
+    HdrHistogram(
+        long maxSnapshotAgeMs,
+        long highestTrackableValue,
+        int numberOfSignificantValueDigits
+    ) {
+        this.maxSnapshotAgeMs = maxSnapshotAgeMs;
+        recorder = new Recorder(highestTrackableValue, numberOfSignificantValueDigits);
+        this.timestampedHistogramSnapshot = new AtomicReference<>(new Timestamped<>(0, null));
+    }
+
+    private Histogram latestHistogram(long now) {
+        Timestamped<Histogram> latest = timestampedHistogramSnapshot.get();
+        while (now - latest.timestamp > maxSnapshotAgeMs) {
+            Histogram currentSnapshot = recorder.getIntervalHistogram();
+            boolean updatedLatest = timestampedHistogramSnapshot.compareAndSet(
+                latest, new Timestamped<>(now, currentSnapshot));
+
+            latest = timestampedHistogramSnapshot.get();
+            if (updatedLatest) {
+                break;
+            }
+        }
+        return latest.value;
+    }
```

Comment:
This part of the implementation differs from the source version in a way that allows concurrent calls that have to update the histogram to lose the state since the last snapshot and update the snapshot with an empty-ish histogram.
E.g.
1. Thread A calls `latestHistogram` and it finds that the current snapshot is outdated. It proceeds to take an interval histogram (which practically resets the internal histogram state), but it gets preempted before it CAS-es it to the snapshot.
2. Thread B calls `latestHistogram` and it finds that the current snapshot is outdated. It proceeds to take an interval histogram (which now contains very little or no data) and successfully CAS-es it to the snapshot.
3. Thread A fails its CAS attempt which used the actual histogram data since the last interval snapshot and ends up using the almost empty snapshot from Thread B.

I'll try to post a PR addressing this soon.

Reply:
That makes sense, please let me know and I will take a look. Thanks!

Reply:
@jeffkbkim Here's the PR, sorry for the slight delay: https://github.com/apache/kafka/pull/17184